### ISSUE TITLE
Could/should SVGGraphics2D dimensions use double precision?

### ISSUE THREAD

**Issue by christianAppl** [jfree/jfreesvg#37]
**TL;DR:**
Why not use double precision for width and height in SVGGraphics2D?

**When applying the matrix:**
PlanarMatrix transformation = new PlanarMatrix().translate(x, y).scale(1.0, -1.0);
To a created svg, I encountered an odd behaviour and that lead to some questions.

The matrix is used to mirror shapes on the y axis in place. Meaning: The x and y translation is used to translate the mirrored shapes back to their original position, after they have been mirrored.
This matrix had originally been defined and applied to paths rendered to BufferedImage objects.

However if this same method is applied to the SVG, then the position will be off.

**The reason is rather obvious:**
My translation assumes that the Graphics dimensions equal the dimensions of the contained path. but the svg paths (as is true for paths in general) does not really need to have integer dimensions. A SVG could easily have a height of 5.678, but the used Graphics would have a height of 6px.

Obviously I have to adapt this to reflect the fact, that the path dimensions and the Graphics dimensions can differ. Easy and not a problem at all.

**Question:**
But what about SVGGraphics2D? Why does it use Integer dimensions in the first place?
(I am fully aware, that I am talking about differences <1 here, and that this might seem to be a non-issue - but the differences are just big enough to be visible and that bothers me. (although I can fix this on my end however.))

**Comment by jfree** [comment]
I don't recall exactly but I suppose my thinking was influenced by:
- most SVG examples on the web use integer width and height;
- my original goal with JFreeSVG was to support the rendering of JFreeChart instances within Swing applications, and here the dimensions always match a panel size which is always specified in integer coordinates;
- switching to double would require some thinking about how many decimal places to write to the SVG output, should scientific notation be supported etc?

I'm not against switching the coordinates to use floating precision, but what are your thoughts on a robust implementation that takes care of the last point above?

**Comment by christianAppl** [comment]
**Scientific notation:**
I really did not know, but I found an answer. CSS seems to support the scientific notation:
![grafik](https://user-images.githubusercontent.com/22076366/111756466-843f1580-889a-11eb-9d7a-d49226071d1b.png)
As SVGs should be using the same number format, this should be okay?

**Precision:**
Using floating point precision clearly has it's issues... I already tested that out and encountered a number of problems, where browser rasterizers reduced precision (by rounding, antialiasing and so forth) and caused gaps and seams, where mathematically none should exist. I can see that my suggestion has flaws.

**I want to derive from my initial question/suggestion:** 
This should not be forced upon a user of this library - I can understand how this would only cause further issues for many use cases, if it was enforced.

However the _option_ to use double precision for width and height would be nice!
I already had a look at you solution for "transformDP" and "geometryDP" why not copy that logic? 
(I Initially had the thought, that I could find a way to improve upon that, but your solution already is absolutely fine, nice and adaptable as it is.)

**Suggestion:**
The type of the fields width and height could be changed to double precision and additionally a "dimensionDP"/"svgDP"/whatever could be introduced.
The dimensionDP would default to 0 (integer precision) and could be set to any value, that is to the users liking.

The only thing concerning me about the solution via those DP fields is the actual handling of the eliminated fractional digits (mostly because currently I don't really know how they are handled.)
Would be nice being able to select how such values shall be treated (FLOOR, CEIL, ROUND) maybe even on a case by case basis. ie: round up to transformDP digits, but always round naturally to geometryDP digits)

**Example using the case at hand:**
Whatever we do, we most likely don't want to cut off paths, so if we also want to use integer precision, we will always prefer providing larger dimensions than required.
A possibly set double width of 68.79 should always be rounded up to the next higher integer (ceil).

Usecase: My path has Rectangle2D.Double bounds and I want to create a SVG with integer precision, that is just big enough to contain it.

**Comment by mhschmieder** [comment]
I would say there's an expectation of the availability of floating-point precision in SVG these days.

I’m just adding some perspective on the overall validity of the original question of supporting the precision, whether optional or default, separate from the caveats and also the limitations of overriding the AWT graphics2D class which has so many integer-based methods.

I did an awful lot of SVG work in my last job (which only ended a few weeks ago), in the JavaFX context vs. AWT/Swing, so I belatedly looked into what was going on behind the scenes in the API calls to see if it is informative towards this AWT context and/or the potential for browser rendering gaps.

In JavaFX, the Shape classes can set String-based content that is presumed to be SVG commands. These in turn get forwarded to the QuantumToolkit which the returns to the JavaFX Path2D class to do basic input parsing of the SVG commands in similar fashion to what we are used to when writing output parsers (the source code for the JavaFX Path2D class is open so is easy to find in the run-time JAR and to look at in an IDE, and I recommend doing so). Everything is declared as single-precision, and my recollection is that this is the SVG standard par the W3C (yes, in Java double-precision is usually MORE efficient).

The precision was absolutely critical for shrunk-down 16x16 vectorized icons in particular. And as the use cases of an SVG output parser in downstream workflows is wide-open, I think it highly likely that a lot of consumers of SVG output from Java AWT (or from JavaFX via AWT using the excellent jfxConverter toolkit that I have also made some contributions to but did not author), will want, need, and/or demand precision.

Speaking from my own experience, in my long-term job before the short-term startup company job, my review committee of end user beta feedback, had me remove SVG support from the app (not the code; just the one line that exposed the menu export method), due to it not being as clean (gaps, integer-slamming, etc.) as PDF, EPS, and other vector output formats. But as SVG is an important part of today’s ecosystem, it is often the only valid choice for a lot of application stacks, and I think it’s important we be as up to date as possible in addressing the downstream client needs.

I have been following this discussion all week but haven’t had time to delve deeply into the issues raised, regarding compatibility vs. previous integer-based assumptions (and potential side effects on existing client code), and technical considerations that have been raised. I am busy 14+ hours a day seven days a week with a job search combined with volunteer work, so unfortunately have not had the time to see what I can do to help resolve some of the conundrums that have come up, but hope to be able to do so over the next few days. For now, I at least wanted to add this additional perspective.

**Comment by mhschmieder** [comment]
Some useful advice in this o'Reilly article on SVG and numeric precision:

https://oreillymedia.github.io/Using_SVG/extras/ch08-precision.html

"For reliable results cross-browser, use numbers with no more than 2 digits after the decimal and four digits before it."

Note that they do mention that scientific notation is supported by the spec (which, as I remembered correctly, is stuck at single-precision as well), as did several other references I found, and also the SVG's that I consumed in my last application development role often had scientific notation as the SVG's were produced by a UX mockup tool and that was how it swang.

**Comment by christianAppl** [comment]
Thank you for your response! I'm not that fluent in all things "web-development" and SVG therefore a more experienced perspective is most welcome!
I already tried to implement dimensions using doubles for SVGGraphics2D and tbh I expected more issues. There is one method in "SVGGraphicsConfiguration" that returns a Rectangle (integer precision) which I solved quick and dirty via:
![grafik](https://user-images.githubusercontent.com/22076366/111958864-0a06cf00-8aee-11eb-9e01-e84e62010d07.png)

As most other methods, that would insist on using integer precision are mostly input methods (as far as I can see) whose parameters are interpreted and translated to the SVG context anyway:
![grafik](https://user-images.githubusercontent.com/22076366/111959254-800b3600-8aee-11eb-85a9-f39d2a4b5b23.png)
(translation to SVG floating point precision in method call "fill()".

Nothing would however stop one, to implement an alternate method of the same name using double (or float) parameters.
This doesn't bother me much however, as for my usecase I mostly require the ability to draw and fill "Shape" objects which already use double precision most of the time.

This ticket is really mostly talking about the fields "width" and "height" of "SVGGraphics2D" which currently are using integer precision where double precision could be used easily without having to change much about the inner workings of the rest of the class:
![grafik](https://user-images.githubusercontent.com/22076366/111960091-95349480-8aef-11eb-98b2-5ba57ddf2b74.png)

I can see however, that alternte double precision draw"Shape" drawImage methods using double precision could be usefull. Which however - is not one of my requirements yet.

**Concerning "thin white lines" in the browser presentation of SVGs**
**This is beyond the scope of this ticket! The SVG has not even been created using this library! This also currently has no priority - I will deal with this and fix it another time. Hints on how to do it are always welcome however. ;)**

Thank you very much for your advice on precision (2 fractional digits)! I already found that recommendation, but currently it unfortunately didn't help me much.
I wrote the SVG manually, that creates the following shape and I am absolutely certain, that those rectangles share identical borders (the constructing points are not _close_ to one another, they are using _identical_ coordinates). Therefore they should construct a seamless plane, but - as you can see - they don't.
I played arround with "shape-rendering" but some settings seem to even worsen the issue.
My assumption is: the SVG is fine, but not browser ready - there is something I am not seeing here:
![grafik](https://user-images.githubusercontent.com/22076366/111961414-530c5280-8af1-11eb-8252-d86fe3aa6fbe.png)
![grafik](https://user-images.githubusercontent.com/22076366/111962083-27d63300-8af2-11eb-9655-3f2be5cc00b9.png)

As far as I am concerned the issue could even be found elsewhere entirely (transform-origin handling of matrices)... I could not reproduce such behaviour using natural numbers - which strongly points to the SVG in my opinion.

**Comment by christianAppl** [comment]
Now I will derail my own ticket entirely - sorry about that!

However - this more or less questions whether or not the intention of this issue is relevant at all:
@mhschmieder Would you recommend using natural numbers for SVGs wherever possible? I read through some of the links you posted and can understand why a perfectionstic view of "I want maximum precision wherever possible" can backfire terribly.

I'm aware, that somewhere along the line the SVG needs to be rastered to pixels, which must always result in "a loss of precision" - as one can't display 0.34 pixels.
- Is the only flaw here my own intention to use the original values?
- Would it be easier and a better idea (without relevant losses) to "flatten" my shape to int precision, before even thinking about creating the SVG?

**TL;DR:**
What is the use of a feature, that no SVG viewer is expected and can be expected to display properly?

**Comment by mhschmieder** [comment]
I won't have time to come back to this until later today, but meant to add a comment last night that the rejection of SVG by my former employer was before I switched to the JFreeOrg library. I did that on my own while unemployed last summer, and it made a big difference. This is my fourth SVG Export solution and definitely my final one, which is why I'm motivated to help where I can.

Just remember that some of my notes about precision are situationally important. For instance, showing every detail of an IFFT (Inverse Finite Fourier Transform, often used to show an Impulse Response), is critical from a data analysis and comprehension point of view vs. an aesthetic basis. And 16x16 down-sized vector graphics icons may be sharper when left to something like a JavaFX rendering engine to properly downscale, if the original larger-sized SVG vector graphics are precise.

**Comment by christianAppl** [comment]
**Concerning this ticket:**
What I can say concerning this issue is. It would be relatively easy to handle the dimensions with double precision instead of int and for some usecases this might simplify the handling of the SVGGraphics2D. (as explained above.)
I agree, that precision should be limited similar to how precision and fractional digits are already limited in the current solution.

If the purpose specifically would work "nicer" with integer dimensions a value of 0 for the maximum fractional digits can be set. Something to specify the handling of eliminated fractional digits would be nice, but it could also be argued, that the user should solve such problems, if he required path "optimization".

I would not opose such a solution and would still claim, that this would be closer to the SVG specification, as SVGs indeed need not have integer dimensions.
To also have said the following:
- It even could be a good idea being able to specifically _not_ set fixed dimensions for the SVG at all, to make it "fully responsive".
- Possibly it even would be a possibility to dynamically calculate width and height for the SVG, by reflecting upon all pathes, that are currently drawn.

As you already stated - whether and if something is a good idea depends highly on the context and the intended purpose of the SVG.

**Concerning my other ramblings:**
I can imagine, that other rasterizers might handle this situation in another way, so that more precise results are possible and/or necessary. Maybe this ticket still has relevance.
But you more or less proved to me, that in my specific usecase "precision" is the cause of my issues and not the solution. (Creating SVGs for usage in html, that is.)
Maybe even concerning that there would be options better fit to display such SVGs in Firefox/Chrome still: From my point of view flattening the shapes seems to be the easiest way to circumvent such issues altogether. I could find a number of suggested workarrounds, that were not really to my liking and that partially didn't even convince me, that they would work. (Adding a 1px border for example.)

Having read some articles and discussions concerning this topic convinced me, that I should rather prepare the values to be fit being displayed without issues, than trying to convince browsers to handle it differently.
There even exist tools to "clean" SVGs to make them "browser ready" which more or less simply round all contained path coordinates up or down.
Therefore I question whether my goal to preserve as much precision as possible was the right approach in the first place.
Superficially - sure... more precision always is nice - but in the end: I want to display shapes, so that they look mostly identical to their original appearance, if I can reach that goal, it does not really matter that much, whether they are exactly and mathematically identical to their source.
If also the required steps for that are simpler, than all that I'm currently doing...

Maybe this issue still has merit - I don't know how fx is handling this tbh.
I know however how Swing/AWT would be handling this (as I know examples for that). And that also points me to: Your SVG is rastered at the end - be ready for that and expect losses.

**Comment by mhschmieder** [comment]
Thanks for the thoughtful response. Now that the semi-annual Oracle Java Update Seminar is over with, I have gone back to my rushed comment from yesterday and improved its clarity plus added an additional point of interest; still situational in nature.

Which just leads me to conclude that the user domain will determine the need, but it should default to what it has been, for the element of least surprise.

And using the RenderingHint mechanism to help isolate the changes (along with floating-point versions of the key drawing commands), should also allow for backward compatibility of results (though this would need testing).

**Comment by jfree** [comment]
Thanks for all the inputs.  I'm going to attempt switching the width and height attributes to double precision.  

One side-note: all the methods that have int parameters and are marked with @Override annotations come from the Java2D Graphics2D API and won't ever be changed in JFreeSVG.  Most of them are, in fact, there to ensure the Graphics2D class maintains compatibility with the older java.awt.Graphics class.  You can generally perform the same operations, but with double precision, using the draw(Shape) and fill(Shape) methods that were introduced by Java2D (back in 1999).

**Comment by mhschmieder** [comment]
Sounds like a good plan, David. And of course this is what I was alluding to when mentioning the integer basis of AWT's Graphics base class earlier. I think I closely followed your own methodology when I did my EPS library over again from scratch last summer. It allowed for a very clean decoupling of integer-vs-float parameter passing and handling, using draw and shape methods as the ;primary "heavy lifters" that then get forwarded to format-specific writers outside the Graphics class hierarchy.

**Comment by jfree** [comment]
Fix will be included in version 5.0
